### PR TITLE
build: move inline resources script to tools

### PR DESCRIPTION
**Pull request by DevVersion**
* Moves the inline resources script to the `tools/gulp/packaging` folder. This is necessary when exposing the build tools (e.g for flex-layout)
* This also makes the packaging more clean because no external script is referenced (also switched to TypeScript and improved code)

### CODE REVIEW

File: tools/gulp/packaging/inline-resources.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
+/* tslint:disable:no-eval */
+
+import {dirname, join} from 'path';
+import {readFileSync, writeFileSync} from 'fs';
+import {sync as glob} from 'glob';
+
+/** Finds all JavaScript files and inlines all external resources of Angular components. */
+export function inlineResourcesFolder(folderPath: string) {
```
Comment (jelbourn) [inline tools/gulp/packaging/inline-resources.ts]: how about just `inlineAllResources`?

File: tools/gulp/packaging/inline-resources.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+  fileContent = inlineTemplate(fileContent, filePath);
+  fileContent = inlineStyles(fileContent, filePath);
+  fileContent = removeModuleId(fileContent);
+
+  writeFileSync(filePath, fileContent, 'utf-8');
+}
+
+/** Inlines the templates of Angular components for a specified source file. */
+function inlineTemplate(fileContent: string, filePath: string) {
```
Comment (jelbourn) [inline tools/gulp/packaging/inline-resources.ts]: Are the inlined resources here minified?

File: tools/gulp/packaging/inline-resources.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    const templateContent = loadResourceFile(templateFile);
+    return `template: "${templateContent}"`;
+  });
+}
+
+
+/** Inlines the external styles of Angular components for a specified source file. */
+function inlineStyles(fileContent: string, filePath: string) {
+  return fileContent.replace(/styleUrls:\s*(\[[\s\S]*?])/gm, (match, styleUrls) => {
+    // The RegExp matches the array of external style files. This is a string right now and
```
Comment (jelbourn) [inline tools/gulp/packaging/inline-resources.ts]: Expand comment to include something like
```ts
// styleUrls will be a string that looks something like "['button.css', 'other.css']"
```

File: tools/gulp/packaging/inline-resources.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    const templateFile = join(dirname(filePath), templateUrl);
+    const templateContent = loadResourceFile(templateFile);
+    return `template: "${templateContent}"`;
+  });
+}
+
+
+/** Inlines the external styles of Angular components for a specified source file. */
+function inlineStyles(fileContent: string, filePath: string) {
+  return fileContent.replace(/styleUrls:\s*(\[[\s\S]*?])/gm, (match, styleUrls) => {
```
Comment (jelbourn) [inline tools/gulp/packaging/inline-resources.ts]: Maybe rename `styleUrls` to `styleUrlsValue` to make it more clear that it's just yet the real array of styleUrl strings. Then `prsedUrls` below could just change to `styleUrls`.

File: tools/gulp/packaging/inline-resources.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    // The RegExp matches the array of external style files. This is a string right now and
+    // can to be parsed using the `eval` method.
+    const parsedUrls = eval(styleUrls) as string[];
+
+    return 'styles: [' + parsedUrls.map(styleUrl => {
+      const stylePath = join(dirname(filePath), styleUrl);
+      const styleContent = loadResourceFile(stylePath);
+      return `"${styleContent}"`;
+    }).join(',\n') + ']';
```
Comment (jelbourn) [inline tools/gulp/packaging/inline-resources.ts]: Maybe refactor this to
```ts
const styles = parsedUrls
    .map(url => join(dirname(filePath), url))
    .map(path => loadResourceFile(path));

return `styles: [${styleContents.join(',')}]`;
```
(the newline shouldn't be necessary AFAICT)

File: tools/gulp/packaging/inline-resources.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  fileContent = removeModuleId(fileContent);
+
+  writeFileSync(filePath, fileContent, 'utf-8');
+}
+
+/** Inlines the templates of Angular components for a specified source file. */
+function inlineTemplate(fileContent: string, filePath: string) {
+  return fileContent.replace(/templateUrl:\s*'([^']+?\.html)'/g, (match, templateUrl) => {
+    const templateFile = join(dirname(filePath), templateUrl);
+    const templateContent = loadResourceFile(templateFile);
```
Comment (jelbourn) [inline tools/gulp/packaging/inline-resources.ts]: Rename variables to `templatePath` and `template`?

File: tools/gulp/packaging/inline-resources.ts (hunk 2 above)
Comment (DevVersion) [inline tools/gulp/packaging/inline-resources.ts]: No, the minifying happens before by Gulp.

File: tools/gulp/packaging/inline-resources.ts (hunk 1 above)
Comment (DevVersion) [inline tools/gulp/packaging/inline-resources.ts]: I thought about that but `AllResources` sounds like it all resources a given file.

File: tools/gulp/packaging/inline-resources.ts (hunk 4 above)
Comment (DevVersion) [inline tools/gulp/packaging/inline-resources.ts]: Yeah I like that.

File: tools/gulp/packaging/inline-resources.ts (hunk 5 above)
Comment (DevVersion) [inline tools/gulp/packaging/inline-resources.ts]: I like that. Not sure about the line break. I didn't write it but I will test it.

File: tools/gulp/packaging/inline-resources.ts (hunk 6 above)
Comment (DevVersion) [inline tools/gulp/packaging/inline-resources.ts]: `templatePaths` sounds good. For the `template` I'd like to keep the `content` to indicate it's the text.

File: tools/gulp/packaging/inline-resources.ts (hunk 1 above)
Comment (jelbourn) [inline tools/gulp/packaging/inline-resources.ts]: Then maybe `inlineAllResourceForDirectory`?

File: tools/gulp/packaging/inline-resources.ts (hunk 6 above)
Comment (jelbourn) [inline tools/gulp/packaging/inline-resources.ts]: That's fine too

File: tools/gulp/packaging/inline-resources.ts (hunk 1 above)
Comment (DevVersion) [inline tools/gulp/packaging/inline-resources.ts]: Yeah `inlineAllResourceForDirectory` sounds good

Comment (DevVersion) [inline tools/gulp/packaging/inline-resources.ts]: Actually why not just `inlineResourcesForDirectory`?

**Comment by DevVersion** [comment]
@jelbourn Addressed your comments. The `inlineStyles` task now looks way better. Thanks

File: tools/gulp/packaging/inline-resources.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    const templateContent = loadResourceFile(templatePath);
+    return `template: "${templateContent}"`;
+  });
+}
+
+/** Inlines the external styles of Angular components for a specified source file. */
+function inlineStyles(fileContent: string, filePath: string) {
+  return fileContent.replace(/styleUrls:\s*(\[[\s\S]*?])/gm, (match, styleUrlsValue) => {
+    // The RegExp matches the array of external style files. This is a string right now and
+    // can to be parsed using the `eval` method. The value looks like "['AAA.css', 'BBB.css"]"
```
Comment (jelbourn) [inline tools/gulp/packaging/inline-resources.ts]: That second-to-last `"` should be a `'`

**Pull request merged**